### PR TITLE
add infrastructure_config part for AWS provider

### PR DESCRIPTION
**Pull request by neo-liang-sap**

### CODE REVIEW

**Comment by neo-liang-sap** [comment]
the PR is verified 
image
image

**Comment by rakesh-garimella** [comment]
Hi @neo-liang-sap  thanks for the PR can you please add tests too. Also we would need changes to flatten_spec too to have support for aws infrastructure config.

**Review by rakesh-garimella — CHANGES_REQUESTED** [review]
Please check the comments.

**Comment by neo-liang-sap** [comment]
> Hi @neo-liang-sap thanks for the PR can you please add tests too. Also we would need changes to flatten_spec too to have support for aws infrastructure config.

Hi @rakesh-garimella :

I added test case for expand_spec and runs ok, see commit https://github.com/kyma-incubator/terraform-provider-gardener/pull/59/commits/110840ce8d6f1bb603e14340615aab9e14540183

One thing i didn't figure out is when i was trying to add flatten_spec and test case for flatten_spec, somehow whatever i tried i failed at test case in handling `zones` for `infrasture_config.network.zones`

The commit for flatten_spec is https://github.com/kyma-incubator/terraform-provider-gardener/pull/59/commits/38d9e9329c2210cf61406154376c695098b2d00a#diff-3964ded6fda0eb3585b3eea32aa61e52R486

The commit for flatten_spec_test is https://github.com/kyma-incubator/terraform-provider-gardener/pull/59/commits/38d9e9329c2210cf61406154376c695098b2d00a#diff-5ee6f244d706b9a81042279823217ed2R355

Basically i referenced `hiberation.schedule` to handle `zones` but test cases reported lots of error which doesn't clear to me, and i didn't googled any valuable information

```
--- FAIL: TestFlattenShootAws (0.00s)
panic: interface conversion: interface {} is *schema.Set, not []interface {} [recovered]
	panic: interface conversion: interface {} is *schema.Set, not []interface {}
```

Do you have any suggestion how i can fix this error?

Thanks for your kind help!

-Neo

**Comment by neo-liang-sap** [comment]
Hi @rakesh-garimella ,

I did some try in fixing the flatten_spec and flatten_spec test, basically the only issue is handling how to `flatten` the value `infrasture_config.network.zones`

Unlike other similar data structure e.g `hibernation.schedules` or `taints`, i have to define `zones` to sth like this
`zones := make([]map[string]interface{}, len(in.Networks.Zones))`

Otherwise the expandProvider will fail - as you can see in my previous comment.

see https://github.com/kyma-incubator/terraform-provider-gardener/pull/59/commits/1a21ef4bd95c80d53eb43c2765cc1c4dcc5f2371#diff-3964ded6fda0eb3585b3eea32aa61e52R487

Could you please confirm whether there's some problem in handling `zones` like this? feel free for any suggestion/modifications.

Thanks!
-Neo

File: expand/expand_spec.go [hunk 1]
```diff
@@ -181,6 +185,89 @@ func getAzureConfig(az []interface{}) *corev1beta1.ProviderConfig {
...
+	obj := awsAlpha1.Networks{}
+	if networks == nil {
+		return obj
+	}
+	in := networks[0].(map[string]interface{})
+
+	if v, ok := in["vpc"].([]interface{}); ok {
+		obj.VPC = getVPC(v)
+	}
+	// if v, ok := in["workers"].(string); ok {
```
Comment (rakesh-garimella) [inline expand/expand_spec.go]: Please remove commented out code.

File: shoot/flatten_spec_test.go [hunk 2]
```diff
@@ -350,3 +351,82 @@ func TestFlattenShootAzure(t *testing.T) {
...
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	d.Set("spec", expected)
```
Comment (rakesh-garimella) [inline shoot/flatten_spec_test.go]: Please check for error here.

File: expand/expand_spec.go (hunk 1 above)
Comment (neo-liang-sap) [reply expand/expand_spec.go]: Thanks! the commented code is removed

File: flatten/flatten_spec.go [hunk 3]
```diff
@@ -457,3 +465,45 @@ func flattenAzure(in azAlpha1.InfrastructureConfig) []interface{} {
 
 	return []interface{}{att}
 }
+
+func flattenAws(in awsAlpha1.InfrastructureConfig) []interface{} {
+	att := make(map[string]interface{})
+	net := make(map[string]interface{})
+	vpc := make(map[string]interface{})
+
+	// if in.EnableECRAccess != nil {
```
Comment (rakesh-garimella) [inline flatten/flatten_spec.go]: We can comment this in I suppose since its used during expand_spec.

**Comment by rakesh-garimella** [comment]
@neo-liang-sap  Can you please add support for control_plane_config too ? Then we can have complete support for aws. Please refer here I have added for gcp: https://github.com/rakesh-garimella/terraform-provider-gardener/tree/gcp-support

basic schema should be similar to infrastructure_config. So 
```
control_plane_config{
aws{
}
}
```

File: flatten/flatten_spec.go (hunk 3 above)
Comment (neo-liang-sap) [reply flatten/flatten_spec.go]: PR updated with latest commit.

**Comment by neo-liang-sap** [comment]
will work on control_plane_config in https://github.com/kyma-incubator/terraform-provider-gardener/issues/64

**Review by rakesh-garimella — APPROVED** [review]

**Review by rakesh-garimella — APPROVED** [review]

**Comment by neo-liang-sap** [comment]
thanks! @rakesh-garimella
